File: celloud/src/main/java/com/celloud/model/mongo/BSI.java
package com.celloud.model.mongo;

import java.util.List;
import java.util.Map;

/**
 * BSI数据报告内容
 * 
 * @author <a href="mailto:liuqingxiao@celloud.cn">liuqx</a>
 * @date 2016年4月14日 下午2:08:14
 * @version Revision: 1.0
 */
public class BSI extends MIB {
    private static final long serialVersionUID = 1L;
    /**
     * 继承MIB baseInfo填写信息不同：
     * 用户填写信息,报告基本信息——姓名[name],性别[sex],年龄[age],科室[department],住院编号[inpatientNo],
     * 床号[bedNo],ID号[idNo],样本种类[sampleType],样本编号[sampleNo],送检医生[doctor],送检日期[
     * inspectionDate],检测人[inspector],复核人[compounder],报告时间[reportDate]
     */
    /**
     * 20种菌的检测结果。包括：species——菌株名，species_zh——菌株中文名，
     * site1——突变位点1信息，site2——突变位点2信息，site3——突变位点3信息，site4——突变位点4信息，
     * unique_reads_num——唯一识别序列数，seq_proportion——序列比重
     * seq1_name——序列1名称，seq1_no——序列1编号，seq1——代表序列1，
     * seq2_name——序列2名称，seq2_no——序列2编号，seq2——代表序列2，
     * seq3_name——序列3名称，seq3_no——序列3编号，seq3——代表序列3，
     * seq4_name——序列4名称，seq4_no——序列4编号，seq4——代表序列4， cross_species——覆盖区域交叉菌
     */
    private List<Map<String, String>> species_20;
    /**
     * 20种菌之外的其他菌检测结果。包括：species——菌株名，species_zh——菌株中文名，
     * site1——突变位点1信息，site2——突变位点2信息，site3——突变位点3信息，site4——突变位点4信息，
     * unique_reads_num——唯一识别序列数，seq1——代表序列1，seq2——代表序列2，seq3——代表序列3，seq4——
     * 代表序列4
     */
    private List<Map<String, String>> species_other;
    /** 人序列总量 */
    private String human_reads;
    /** 16S序列数 */
    private String reads_16s;
    public List<Map<String, String>> getSpecies_20() {
        return species_20;
    }

    public void setSpecies_20(List<Map<String, String>> species_20) {
        this.species_20 = species_20;
    }

    public List<Map<String, String>> getSpecies_other() {
        return species_other;
    }

    public void setSpecies_other(List<Map<String, String>> species_other) {
        this.species_other = species_other;
    }

    public String getHuman_reads() {
        return human_reads;
    }

    public void setHuman_reads(String human_reads) {
        this.human_reads = human_reads;
    }

    public String getReads_16s() {
        return reads_16s;
    }

    public void setReads_16s(String reads_16s) {
        this.reads_16s = reads_16s;
    }
}

File: python/app/BSI.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

__des__ = "BSI的操作类"
__author__ = "leamox"

import threading
from mongo.mongoOperate import mongo
from utils.StringUtils import *
from utils.FileUtils import *
from app.MIB import MIB

class BSI:
    path = None
    instance = None
    locker = threading.Lock()
    # 初始化
    def __init__(self):
        pass

    # 获取单例对象
    @staticmethod
    def getInstance():
        BSI.locker.acquire()
        try:
            if not BSI.instance:
                BSI.instance = BSI()
            return BSI.instance
        finally:
            BSI.locker.release()

    # 执行
    def getResult(self, path, appId, dataKey):
        result = {}
        result1 = MIB.getInstance().getResult(path, appId, dataKey)
        result = dict(result, **result1)
        path = os.path.join(path, dataKey)
        if not os.path.exists(path):
            return result
        result_path = os.path.join(path, 'result')
        normal_species_path = os.path.join(result_path, "20.species")
        mo = mongo.getInstance()
        if os.path.exists(normal_species_path):
            f = open(normal_species_path, "r")
            species_20 = []
            for line in f.readlines():
                st = line.strip().split("\t")
                specie_name = list_value(st, 0)
                cross_species = list_value(st, 6)
                strain_dict = mo.get({"species": str(specie_name)}, "BSISpecies")
                species_zh = ""
                if strain_dict is not None:
                    species_zh = dict_value(strain_dict, "species_zh")
                cross_species_dict = mo.get({"species": str(cross_species)}, "BSISpecies")
                cross_species_zh = ""
                if cross_species_dict is not None:
                    cross_species_zh = dict_value(cross_species_dict, "species_zh")
                seq1 = list_value(st, 7)
                seq1_name = ""
                seq1_no = ""
                seq1_text = ""
                if seq1 is not None:
                    seq1list = seq1.split(" ")
                    seq1_name = list_value(seq1list, 0)
                    seq1_no = list_value(seq1list, 1)
                    seq1_text = list_value(seq1list, 2)
                seq2 = list_value(st, 8)
                seq2_name = ""
                seq2_no = ""
                seq2_text = ""
                if seq2 is not None:
                    seq2list = seq2.split(" ")
                    seq2_name = list_value(seq2list, 0)
                    seq2_no = list_value(seq2list, 1)
                    seq2_text = list_value(seq2list, 2)
                seq3 = list_value(st, 9)
                seq3_name = ""
                seq3_no = ""
                seq3_text = ""
                if seq3 is not None:
                    seq3list = seq3.split(" ")
                    seq3_name = list_value(seq3list, 0)
                    seq3_no = list_value(seq3list, 1)
                    seq3_text = list_value(seq3list, 2)
                seq4 = list_value(st, 10)
                seq4_name = ""
                seq4_no = ""
                seq4_text = ""
                if seq4 is not None:
                    seq4list = seq4.split(" ")
                    seq4_name = list_value(seq4list, 0)
                    seq4_no = list_value(seq4list, 1)
                    seq4_text = list_value(seq4list, 2)
                species = {"species": specie_name, "species_zh": species_zh,
                           "site1": list_value(st, 1), "site2": list_value(st, 2),
                           "site3": list_value(st, 3), "site4": list_value(st, 4),
                           "seq_proportion": list_value(st, 5),
                           "seq1_name": seq1_name, "seq1_no": seq1_no, "seq1": seq1_text,
                           "seq2_name": seq2_name, "seq2_no": seq2_no, "seq2": seq2_text,
                           "seq3_name": seq3_name, "seq3_no": seq3_no, "seq3": seq3_text,
                           "seq4_name": seq4_name, "seq4_no": seq4_no, "seq4": seq4_text,
                           "cross_species": cross_species,"cross_species_zh": cross_species_zh
                           }
                species_20.append(species)
            result["species_20"] = species_20
            f.close()
        other_species_path = os.path.join(result_path, "other.species")
        if os.path.exists(other_species_path):
            f = open(other_species_path, "r")
            species_other = []
            for line in f.readlines():
                st = line.strip().split("\t")
                species = {"species": list_value(st, 0), "site1": list_value(st, 1),
                           "site2": list_value(st, 2), "site3": list_value(st, 3), "site4": list_value(st, 4),
                           "unique_reads_num": list_value(st, 5), "seq1": list_value(st, 7), "seq2": list_value(st, 8),
                           "seq3": list_value(st, 9), "seq4": list_value(st, 10)
                           }
                species_other.append(species)
            result["species_other"] = species_other
            f.close()
        return result



